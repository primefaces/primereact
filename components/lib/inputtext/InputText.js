import React, { forwardRef, memo, useEffect, useMemo, useRef } from 'react';
import PropTypes from 'prop-types';
import { KeyFilter } from '../keyfilter/KeyFilter';
import { tip } from '../tooltip/Tooltip';
import { DomHandler, ObjectUtils, classNames } from '../utils/Utils';
import { useUnmountEffect } from '../hooks/Hooks';

export const InputText = memo(forwardRef((props, ref) => {
    const elementRef = useRef(ref);
    const tooltipRef = useRef(null);

    const onKeyPress = (event) => {
        props.onKeyPress && props.onKeyPress(event);

        if (props.keyfilter) {
            KeyFilter.onKeyPress(event, props.keyfilter, props.validateOnly)
        }
    }

    const onInput = (event) => {
        let validatePattern = true;
        if (props.keyfilter && props.validateOnly) {
            validatePattern = KeyFilter.validate(event, props.keyfilter);
        }

        props.onInput && props.onInput(event, validatePattern);

        if (!props.onChange) {
            const target = event.target;
            ObjectUtils.isNotEmpty(target.value) ? DomHandler.addClass(target, 'p-filled') : DomHandler.removeClass(target, 'p-filled');
        }
    }

    const isFilled = useMemo(() => (
        ObjectUtils.isNotEmpty(props.value) || ObjectUtils.isNotEmpty(props.defaultValue) || (elementRef.current && ObjectUtils.isNotEmpty(elementRef.current.value))
    ), [props.value, props.defaultValue]);

    useEffect(() => {
        ObjectUtils.combinedRefs(elementRef, ref);
    }, [elementRef, ref]);

    useEffect(() => {
        if (tooltipRef.current) {
            tooltipRef.current.update({ content: props.tooltip, ...(props.tooltipOptions || {}) });
        }
        else if (props.tooltip) {
            tooltipRef.current = tip({
                target: elementRef.current,
                content: props.tooltip,
                options: props.tooltipOptions
            });
        }
    }, [props.tooltip, props.tooltipOptions]);

    useUnmountEffect(() => {
        if (tooltipRef.current) {
            tooltipRef.current.destroy();
            tooltipRef.current = null;
        }
    });

    const inputProps = ObjectUtils.findDiffKeys(props, InputText.defaultProps);
    const className = classNames('p-inputtext p-component', {
        'p-disabled': props.disabled,
        'p-filled': isFilled
    }, props.className);

    return <input ref={elementRef} {...inputProps} className={className} onInput={onInput} onKeyPress={onKeyPress} />;
}));

InputText.defaultProps = {
    __TYPE: 'InputText',
    keyfilter: null,
    validateOnly: false,
    tooltip: null,
    tooltipOptions: null,
    onInput: null,
    onKeyPress: null
}

InputText.propTypes /* remove-proptypes */ = {
    __TYPE: PropTypes.string,
    keyfilter: PropTypes.any,
    validateOnly: PropTypes.bool,
    tooltip: PropTypes.string,
    tooltipOptions: PropTypes.object,
    onInput: PropTypes.func,
    onKeyPress: PropTypes.func
}
