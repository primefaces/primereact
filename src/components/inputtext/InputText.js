import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export class InputText extends Component {

    static defaultProps = {
        onInput: null
    };

    static propTypes = {
        onInput: PropTypes.func
    };

    constructor(props) {
        super(props);
        this.onClick = this.onClick.bind(this);
        this.onInput = this.onInput.bind(this);
    }

    onInput(e) {
        e.stopPropagation();
        console.log('onInput');
        if(this.props.onInput) {
            this.props.onInput(e);
        }

        this.updateFilledState(e.target.value);
    }

  onClick(e) {
    if(this.props.onClick) {
      this.props.onClick();
    } else {
      e.stopPropagation();
    }
  }

    updateFilledState(val) {
        if(val && val.length)
            this.inputEl.classList.add('ui-state-filled');
        else
            this.inputEl.classList.remove('ui-state-filled');
    }
    
    componentDidMount() {
        let _value =  this.props.value||this.props.defaultValue;
        
        this.updateFilledState(_value);
    }

    render() {
        var className = classNames('ui-inputtext ui-state-default ui-corner-all ui-widget', this.props.className, {
            'ui-state-disabled': this.props.disabled
        });

        let inputProps = Object.assign({}, this.props);
        delete inputProps.onInput;

        return <input ref={(el) => this.inputEl = el} {...inputProps} className={className} onInput={this.onInput} onClick={this.onClick}/>;
    }
}