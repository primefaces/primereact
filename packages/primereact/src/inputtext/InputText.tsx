'use client';
import { Component } from '@primereact/core/component';
import { useInputText } from '@primereact/headless/inputtext';
import { styles } from '@primereact/styles/inputtext';
import { mergeProps } from '@primeuix/utils';
import { withComponent } from 'primereact/base';
import { useFluidContext } from 'primereact/fluid';
import * as React from 'react';
import { InputTextProvider } from './InputText.context';
import { defaultProps } from './InputText.props';

export const InputText = withComponent({
    name: 'InputText',
    defaultProps,
    styles,
    setup(instance) {
        const inputtext = useInputText(instance?.inProps);
        const fluid = useFluidContext();

        return { ...inputtext, fluid };
    },
    render(instance) {
        const { id, props, ptmi, cx, fluid } = instance;

        const rootProps = mergeProps(
            {
                id,
                type: 'text',
                className: cx('root', { $fluid: !!fluid })
            },
            ptmi('root')
        );

        return (
            <InputTextProvider value={instance}>
                <Component instance={instance} attrs={rootProps} children={props.children} />
            </InputTextProvider>
        );
    }
});
