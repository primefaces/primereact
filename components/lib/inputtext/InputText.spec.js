import '@testing-library/jest-dom';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as React from 'react';
import { InputText } from './InputText';

describe('InputText', () => {
    test('when input is enabled then input accepts data entry and have filled state', () => {
        // Arrange
        const { container } = render(<InputText />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        fireEvent.input(input, { target: { value: 'abc' } });

        // Act
        expect(input).toBeEnabled();
        expect(input).toHaveValue('abc');
        expect(container).toMatchSnapshot();
    });
    test('when input is blank it should not have filled state', () => {
        // Arrange
        const { container } = render(<InputText disabled={false} />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        fireEvent.input(input, { target: { value: '' } });

        // Act
        expect(input).toBeEnabled();
        expect(input).toHaveValue('');
        expect(container).toMatchSnapshot();
    });
    test('when input is is set for validation only', () => {
        // Arrange
        const { container } = render(<InputText validateOnly keyfilter={`alpha`} />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        fireEvent.input(input, { target: { value: 'def' } });

        // Act
        expect(container).toMatchSnapshot();
        expect(input).toHaveValue('def');
    });
    test('when input is disabled it should render as disabled', async () => {
        // Arrange
        const { container } = render(<InputText disabled />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        await userEvent.type(input, '123');

        // Act
        expect(input).toBeDisabled();
        expect(input).toHaveValue('');
        expect(container).toMatchSnapshot();
    });
    test('when input is using keyfilter for integers accept integer input', async () => {
        // Arrange
        const keydownOn = jest.fn();
        const { container } = render(<InputText keyfilter="int" onKeyDown={keydownOn} />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        await userEvent.type(input, '123');

        // Act
        expect(input).toHaveValue('123');
        expect(keydownOn).toHaveBeenCalledTimes(3);
    });
    test('when input is using keyfilter for integers do not accept alphabetic input', async () => {
        // Arrange
        const keydownOn = jest.fn();
        const { container } = render(<InputText keyfilter="int" onKeyDown={keydownOn} />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        await userEvent.type(input, 'abc');

        // Act
        expect(input).toHaveValue('');
        expect(keydownOn).toHaveBeenCalledTimes(3);
    });
    test('when input is using keyfilter for alphabetic accept paste of alphabetic values', async () => {
        // Arrange
        const pasteOn = jest.fn();
        const { container } = render(<InputText keyfilter="alpha" onPaste={pasteOn} />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        input.focus();
        await userEvent.paste('abc');

        // Act
        expect(input).toHaveValue('abc');
        expect(pasteOn).toHaveBeenCalledTimes(1);
    });
    test('when input is using keyfilter for alphabetic do not accept paste of integer values', async () => {
        // Arrange
        const pasteOn = jest.fn();
        const { container } = render(<InputText keyfilter="alpha" onPaste={pasteOn} />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        input.focus();
        await userEvent.paste('123');

        // Act
        expect(input).toHaveValue('');
        expect(pasteOn).toHaveBeenCalledTimes(1);
    });
    test('when input is using onInput make sure it is called', async () => {
        // Arrange
        const inputOn = jest.fn();
        const ref = React.createRef();
        const { container } = render(<InputText ref={ref} onInput={inputOn} />);
        const input = container.getElementsByTagName('input')[0];

        // Act
        ref.current.value = 'Jest testing value';
        input.focus();
        await userEvent.paste(' abc');

        // Assert
        expect(input).toHaveValue('Jest testing value abc');
        expect(inputOn).toHaveBeenCalledTimes(1);
    });
});
