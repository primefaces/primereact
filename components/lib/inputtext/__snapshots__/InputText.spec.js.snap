// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText when input is blank it should not have filled state 1`] = `
<div>
  <input
    class="p-inputtext p-component"
  />
</div>
`;

exports[`InputText when input is disabled it should render as disabled 1`] = `
<div>
  <input
    class="p-inputtext p-component p-disabled"
    disabled=""
  />
</div>
`;

exports[`InputText when input is enabled then input accepts data entry and have filled state 1`] = `
<div>
  <input
    class="p-inputtext p-component p-filled"
  />
</div>
`;

exports[`InputText when input is is set for validation only 1`] = `
<div>
  <input
    class="p-inputtext p-component p-filled"
  />
</div>
`;
